File: dsa-problems/Recursion/solutions/cpp/REC-004-exam-seating-backtrack.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    int N;
    int count;
    vector<bool> cols;
    vector<bool> diag1;
    vector<bool> diag2;

public:
    int countNQueens(int n) {
        N = n;
        count = 0;
        cols.assign(2 * n, false);
        diag1.assign(2 * n, false);
        diag2.assign(2 * n, false);
        backtrack(0);
        return count;
    }

    void backtrack(int row) {
        if (row == N) {
            count++;
            return;
        }

        for (int col = 0; col < N; col++) {
            if (cols[col] || diag1[row + col] || diag2[row - col + N]) continue;
            
            cols[col] = true;
            diag1[row + col] = true;
            diag2[row - col + N] = true;
            
            backtrack(row + 1);
            
            cols[col] = false;
            diag1[row + col] = false;
            diag2[row - col + N] = false;
        }
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    int n; 
    if (!(cin >> n)) return 0;
    
    Solution sol;
    cout << sol.countNQueens(n) << endl;
    return 0;
}

File: dsa-problems/Recursion/solutions/python/REC-004-exam-seating-backtrack.py
def count_nqueens(n: int) -> int:
    """Count the number of ways to place n queens on an n×n chessboard.
    Uses bitwise operations and memoization for optimization.
    """
    memo = {}

    def backtrack(row, cols, diag1, diag2):
        # Memoization key
        key = (row, cols, diag1, diag2)
        if key in memo:
            return memo[key]

        # Base case: all queens placed
        if row == n:
            return 1

        count = 0
        # Try placing queen in each valid column
        for col in range(n):
            # Check if column and diagonals are available using bitwise AND
            if not (cols & (1 << col)):
                d1 = row - col + n  # Offset to make positive
                d2 = row + col

                if not (diag1 & (1 << d1)) and not (diag2 & (1 << d2)):
                    # Place queen and recurse
                    count += backtrack(
                        row + 1,
                        cols | (1 << col),
                        diag1 | (1 << d1),
                        diag2 | (1 << d2)
                    )

        memo[key] = count
        return count

    return backtrack(0, 0, 0, 0)

def main():
    import sys
    n = int(sys.stdin.read().strip())
    result = count_nqueens(n)
    print(result)

if __name__ == "__main__":
    main()
